
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Modeling &#8212; Data Science Handbook</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-svg.js"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Topology Inference" href="31-topology-inference.html" />
    <link rel="prev" title="Sampling and Estimation" href="13-sampling-and-estimation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      <h1 class="site-logo" id="site-title">Data Science Handbook</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../index.html">
   Data Science Handbook
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Foundations
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../11-math/00-math.html">
   Math
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/11-combinatorics.html">
     Combinatorics
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/20-vector-spaces.html">
     Vector Spaces
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/31-geometry.html">
     Geometry
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/21-linear-algebra.html">
     Linear Algebra
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/51-linear-programming.html">
     Linear Programming
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/52-non-linear-programming.html">
     Non-linear Programming
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../11-math/90-puzzles.html">
     Puzzles
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../12-probabilities/00-probabilities.html">
   Probabilities
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/11-expectation-and-variance.html">
     Expectation and Variance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/13-correlation-and-dependence.html">
     Correlation and Dependence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/31-bayesian-theorem.html">
     Bayesian’s Theorem
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/51-markov-chain.html">
     Markov Chain
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/71-sampling.html">
     Sampling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/90-multivariate-notations.html">
     Multivariate Notations
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/91-exponential-families.html">
     Exponential Families
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../12-probabilities/91-large-sample-theory.html">
     Large Sample Theory
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../13-statistics/00-statistics.html">
   Statistics
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/11-sample-survey.html">
     Sample Survey
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/13-randomized-trial.html">
     Randomized Controlled Trials
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/21-hypothesis-testing.html">
     Hypothesis Testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/23-common-tests.html">
     Common Tests
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/33-confusion-matrix.html">
     Confusion Matrix
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/41-maximum-likelihood-estimation.html">
     Maximum Likelihood Estimator
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../13-statistics/43-estimators-evaluation.html">
     Estimators Evaluation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../15-tools/00-tools.html">
   Tools
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../15-tools/11-python.html">
     Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../15-tools/21-r.html">
     R
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../15-tools/31-sql.html">
     SQL
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../15-tools/41-latex.html">
     LaTeX
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../15-tools/51-myst.html">
     MyST Markdown
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Algorithms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../20-algorithms-concepts/00-algorithms-concepts.html">
   Algorithms Concepts
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../20-algorithms-concepts/51-polynomial-reduction.html">
     Polynomial Reduction
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../20-algorithms-concepts/53-P-and-NP.html">
     <span class="math notranslate nohighlight">
      \(P\)
     </span>
     and
     <span class="math notranslate nohighlight">
      \(NP\)
     </span>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../20-algorithms-concepts/61-randomized-algo.html">
     Randomized Algorithms
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../20-algorithms-concepts/71-streaming.html">
     Streaming Algorithms
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../21-greedy-algorithms/00-greedy-algorithms.html">
   Greedy Algorithms
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../21-greedy-algorithms/11-interval-scheduling.html">
     Interval Scheduling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../21-greedy-algorithms/31-huffman-coding.html">
     Huffman Coding
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../23-dynamic-programming/00-dynamic-programming.html">
   Dynamic Programming
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../23-dynamic-programming/11-weighted-interval-scheduling.html">
     Weighted Interval Scheduling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../23-dynamic-programming/13-longest-common-subsequence.html">
     Longest Common Subsequence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../23-dynamic-programming/15-longest-increasing-subsequence.html">
     Longest Increasing Subsequence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../23-dynamic-programming/17-largest-sum-subsequence.html">
     Largest Sum Subsequence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../23-dynamic-programming/31-knapsack.html">
     Minimum Knapsack
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../23-dynamic-programming/51-chain-matrix-multiplication.html">
     Chain Matrix Multiplication
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../25-graph-related/00-graph-related.html">
   Graph Related
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../25-graph-related/13-shortest-path.html">
     Shortest Path
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../25-graph-related/21-minimum-spanning-tree.html">
     Minimum Spanning Tree
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../25-graph-related/31-maximum-flow.html">
     Maximum Flow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../25-graph-related/32-matching.html">
     Matching
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../25-graph-related/42-maximum-independent-set.html">
     Maximum Independent Set in Trees
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../25-graph-related/91-LP-max-flow-min-cut.html">
     LP on Max-flow and Min-cut
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Machine Learning
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../30-ml-basics/00-ml-basics.html">
   Machine Learning Basics
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/02-taxonomy.html">
     Taxonomy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/03-information-theory.html">
     Information Theory
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/05-kernels.html">
     Kernels
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/11-data-issues.html">
     Data Issues
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/51-semi-supervised.html">
     Semi-supervised Learning
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/53-self-supervised.html">
     Self-supervised Learning
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../30-ml-basics/61-fourier-transform.html">
     Fourier Transform-based Representations
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../31-regression/00-regression.html">
   Regression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/11-lm-estimation.html">
     Linear Models - Estimation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/12-lm-inference.html">
     Linear Models - Inference
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/13-lm-diagnosis.html">
     Linear Models - Diagnosis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/14-lm-advanced.html">
     Linear Models - Advanced Topics
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/21-generalized-linear-models.html">
     Generalized Linear Models
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/22-logistic-regression.html">
     Logistic Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/23-multinomial-logitsitc.html">
     Multinomial Logistic Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/24-ordinal-logistic.html">
     Ordinal Logistic Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/25-poisson-regression.html">
     Poisson Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../31-regression/31-multivariate-regression.html">
     Multivariate Regression
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../32-classification/00-classification.html">
   Classification
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../32-classification/09-k-nearest-neighbors.html">
     K-nearest neighbors
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../32-classification/11-normal.html">
     For Gaussian Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../32-classification/13-linear-discriminant-analysis.html">
     Linear Discriminant Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../32-classification/21-support-vector-machine.html">
     Support Vector Machine
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../32-classification/31-decision-tree.html">
     Decision Tree
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../33-dimensionality-reduction/00-dimensionality-reduction.html">
   Dimensionality Reduction
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/11-principal-component-analysis.html">
     Principal Component Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/13-canonical-correlation-analysis.html">
     Canonical Correlation Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/21-multidimensional-scaling.html">
     Multidimensional Scaling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/23-graph-based-spectral-methods.html">
     Graph-based Spectral Methods
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/25-t-SNE.html">
     SNE and
     <span class="math notranslate nohighlight">
      \(t\)
     </span>
     -SNE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/31-kernel-pca.html">
     Kernel PCA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/32-kernel-cca.html">
     Kernel CCA
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/41-factor-analysis.html">
     Factor Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../33-dimensionality-reduction/51-correspondence-analysis.html">
     Correspondence Analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../34-clustering/00-clustering.html">
   Clustering
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
  <label for="toctree-checkbox-13">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../34-clustering/11-k-means.html">
     <span class="math notranslate nohighlight">
      \(k\)
     </span>
     -means clustering
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../34-clustering/13-agglomerative-methods.html">
     Agglomerative Methods
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../34-clustering/31-spectral-clustering.html">
     Spectral Clustering
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../34-clustering/41-gaussian-mixtures.html">
     Gaussian Mixtures
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../35-graphical-models/00-graphical-models.html">
   Graphical Models
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
  <label for="toctree-checkbox-14">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../35-graphical-models/03-random-walks.html">
     Random Walks in Graphs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../35-graphical-models/11-hidden-markov-models.html">
     Hidden Markov Models
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../35-graphical-models/31-topic-models.html">
     Topic Models
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../35-graphical-models/33-language-models.html">
     Language Models
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../37-neural-networks/00-neural-networks.html">
   Neural Networks
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
  <label for="toctree-checkbox-15">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/01-stochastic-gradient-descent.html">
     Stochastic Gradient Descent
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/03-trainability.html">
     Trainability
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/05-regularization.html">
     Regularization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/11-autoencoders.html">
     Autoencoders
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/13-variational-autoencoders.html">
     Variational Autoencoders
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/31-sequential-models.html">
     Sequential Models
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../37-neural-networks/41-GAN.html">
     Generative Adversarial Networks
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="00-ml-for-graph-data.html">
   For Graph-structured Data
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
  <label for="toctree-checkbox-16">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="01-graph-basics.html">
     Graph Basics
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="11-descriptive-analysis.html">
     Descriptive Analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="13-sampling-and-estimation.html">
     Sampling and Estimation
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Modeling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="31-topology-inference.html">
     Topology Inference
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="41-processes.html">
     Processes on Graphs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="51-graph-rep-learning.html">
     Graph Representation Learning
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/38-ml-for-graph-data/21-modeling.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/dennissxz/data-science-handbook"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/dennissxz/data-science-handbook/issues/new?title=Issue%20on%20page%20%2F38-ml-for-graph-data/21-modeling.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#random-graph-models">
   Random Graph Models
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#background">
     Background
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#estimate-eta-g">
       Estimate
       <span class="math notranslate nohighlight">
        \(\eta(G)\)
       </span>
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#assess-significance-of-eta-g-obs">
       Assess Significance of
       <span class="math notranslate nohighlight">
        \(\eta(G^{obs})\)
       </span>
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#classical-random-graph-models">
     Classical Random Graph Models
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#erdos-and-renyi">
       Erdos and Renyi
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#gilbert">
       Gilbert
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generalized-random-graph-models">
     Generalized Random Graph Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#simulation">
     Simulation
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#classical-rgm">
       Classical RGM
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#generalized-rgm">
       Generalized RGM
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#application">
     Application
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#hidden-population-size">
       Hidden Population Size
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#assessing-significance">
       Assessing Significance
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#detecting-network-motifs">
       Detecting Network Motifs
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#real-world-models">
   Real-world Models
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#small-world-models">
     Small-World Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#kronecker-graph-model">
     Kronecker Graph Model
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#growth-models">
   Growth Models
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#preferential-attachment">
     Preferential Attachment
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#copying-models">
     Copying Models
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#full-copying">
       Full Copying
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#partial-copying">
       Partial Copying
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fitting">
     Fitting
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#exponential-random-graph-models">
   Exponential Random Graph Models
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#bernoulli-random-graphs">
     Bernoulli Random Graphs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#markov-random-graphs">
     Markov Random Graphs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Fitting
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#approximate-log-likelihood">
       Approximate Log-likelihood
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#log-pseudo-likelihood">
       Log-pseudo-likelihood
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#goodness-of-fit">
     Goodness-of-fit
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#model-degeneracy">
     Model Degeneracy
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#challenges">
   Challenges
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="modeling">
<h1>Modeling<a class="headerlink" href="#modeling" title="Permalink to this headline">¶</a></h1>
<div class="section" id="random-graph-models">
<h2>Random Graph Models<a class="headerlink" href="#random-graph-models" title="Permalink to this headline">¶</a></h2>
<div class="section" id="background">
<h3>Background<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h3>
<p>By a model for a graph we mean a collection</p>
<div class="math notranslate nohighlight">
\[
\left\{ \mathbb{P} _\theta (G), G \in \mathcal{G}: \theta \in \Theta  \right\}
\]</div>
<p>where</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\mathcal{G}\)</span> is a collection (‘ensemble’) of possible graphs</p></li>
<li><p><span class="math notranslate nohighlight">\(\mathbb{P}_\theta\)</span> is a probability distribution on <span class="math notranslate nohighlight">\(\mathcal{G}\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\theta\)</span> is a vector of parameters ranging over values in <span class="math notranslate nohighlight">\(\Theta\)</span>.</p></li>
</ul>
<div class="section" id="estimate-eta-g">
<h4>Estimate <span class="math notranslate nohighlight">\(\eta(G)\)</span><a class="headerlink" href="#estimate-eta-g" title="Permalink to this headline">¶</a></h4>
<p>In traditional statistical sampling theory, there are two main approaches to constructing estimates of population parameters <span class="math notranslate nohighlight">\(\eta(G)\)</span> from a sample <span class="math notranslate nohighlight">\(G^*\)</span>: design-based and model-based.</p>
<ul class="simple">
<li><p>design-based: inference is based entirely on the random mechanism by which
a subset of elements were selected from the population to create the sample. We have see examples in the previous section.</p></li>
<li><p>model-based approach: on the other hand, a model is given to specify a relationship between the sample and the population. Model-based estimation strategies including least-squares, method-of-moments, maximum-likelihood, etc are then used for constructing estimators for <span class="math notranslate nohighlight">\(\eta(G)\)</span>.</p></li>
</ul>
<p>In more recent decades, the distinction between these two approaches has become more blurred.</p>
</div>
<div class="section" id="assess-significance-of-eta-g-obs">
<h4>Assess Significance of <span class="math notranslate nohighlight">\(\eta(G^{obs})\)</span><a class="headerlink" href="#assess-significance-of-eta-g-obs" title="Permalink to this headline">¶</a></h4>
<p>Suppose that we have a graph <span class="math notranslate nohighlight">\(G^{obs}\)</span> derived from observations of some sort (i.e., <strong>not</strong> necessarily through a formal network sampling mechanism). We often interested in whether <span class="math notranslate nohighlight">\(\eta(G^{obs})\)</span> is ‘significant’, in the sense that unusual or unexpected.</p>
<p>To measure this, we need a reference, like a ‘null hypothesis’ in hypothesis testing. A RGM can be used to create a reference distribution which, under the accompanying assumption of uniform likelihood of elements in <span class="math notranslate nohighlight">\(\mathcal{G}\)</span>, takes the form,</p>
<div class="math notranslate nohighlight">
\[
\mathbb{P}_{\eta, \mathcal{G}} (t)  = \frac{\# \left\{ G \in \mathcal{G}: \eta(G) \le t \right\}}{\left\vert \mathcal{G} \right\vert}
\]</div>
<p>If <span class="math notranslate nohighlight">\(\eta(G^{obs})\)</span> is found to be sufficiently unlikely under this distribution, this is taken as evidence <strong>against</strong> the hypothesis that Gobs is a uniform draw from <span class="math notranslate nohighlight">\(G\)</span>.</p>
<p>Some issues:</p>
<ul class="simple">
<li><p>How to choose <span class="math notranslate nohighlight">\(\mathcal{G}\)</span>?</p></li>
<li><p>Usually it is not possible to enumerate all elements in <span class="math notranslate nohighlight">\(\mathcal{G}\)</span>, hence, cannot compute <span class="math notranslate nohighlight">\(\mathbb{P}_{\eta, \mathcal{G}} (t)\)</span> exactly <span class="math notranslate nohighlight">\(\rightarrow\)</span> sol: approximation.</p></li>
</ul>
</div>
</div>
<div class="section" id="classical-random-graph-models">
<h3>Classical Random Graph Models<a class="headerlink" href="#classical-random-graph-models" title="Permalink to this headline">¶</a></h3>
<div class="section" id="erdos-and-renyi">
<span id="er-random-graph"></span><h4>Erdos and Renyi<a class="headerlink" href="#erdos-and-renyi" title="Permalink to this headline">¶</a></h4>
<p>Equal probability on all graphs of a given order and size:</p>
<div class="math notranslate nohighlight">
\[\mathcal{G} (N_v, N_e) = \left\{ G = (V, E): \left\vert V \right\vert = N_v, \left\vert E \right\vert = N_e\right\}\]</div>
<p>It is easy to find <span class="math notranslate nohighlight">\(\left\vert \mathcal{G} (N_v, N_e) \right\vert = \binom{\binom{N_v}{2}}{N_e}\)</span>, hence</p>
<div class="math notranslate nohighlight">
\[\mathbb{P} (G) = \binom{\binom{N_v}{2}}{N_e} ^{-1}\]</div>
</div>
<div class="section" id="gilbert">
<h4>Gilbert<a class="headerlink" href="#gilbert" title="Permalink to this headline">¶</a></h4>
<p>A collection <span class="math notranslate nohighlight">\(\mathcal{G} (N_v, p)\)</span> is defined to consist of all graphs <span class="math notranslate nohighlight">\(G\)</span> of order <span class="math notranslate nohighlight">\(N_v\)</span> that may be obtained by assigning an edge <strong>independently</strong> to each pair of distinct vertices with probability <span class="math notranslate nohighlight">\(p\)</span>.</p>
<div class="math notranslate nohighlight">
\[\mathcal{G} (N_v, p) = \left\{ G = (V, E): \left\vert V \right\vert = N_v, \left\vert E \right\vert = N_e\right\}\]</div>
<p>The level of connectivity is related to the relation between <span class="math notranslate nohighlight">\(p\)</span> and <span class="math notranslate nohighlight">\(N_v\)</span>. Let <span class="math notranslate nohighlight">\(p = \frac{c}{N}\)</span> for <span class="math notranslate nohighlight">\(c &gt; 0\)</span>, then</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(c &lt; 1\)</span>: w.h.p. all components will have <span class="math notranslate nohighlight">\(\mathcal{O} (\log N_v)\)</span> vertices. The largest component is a tree. No circles in the graph.</p></li>
<li><p><span class="math notranslate nohighlight">\(c &gt; 1\)</span>: w.h.p. <span class="math notranslate nohighlight">\(G\)</span> will have a single connected component (‘giant component’) consisting of <span class="math notranslate nohighlight">\(\alpha_c N_v\)</span> vertices, for some constant <span class="math notranslate nohighlight">\(\alpha_c &gt; 0\)</span>, with the remaining components having only on the order of <span class="math notranslate nohighlight">\(\mathcal{O} (\log N_v)\)</span> vertices.</p></li>
<li><p><span class="math notranslate nohighlight">\(c &gt; \log n\)</span>: w.h.p. <span class="math notranslate nohighlight">\(G\)</span> will be connected.</p></li>
</ul>
<div class="figure align-default" id="graph-random-transitions">
<a class="reference internal image-reference" href="../_images/graph-random-transitions.png"><img alt="" src="../_images/graph-random-transitions.png" style="width: 30%;" /></a>
<p class="caption"><span class="caption-number">Fig. 193 </span><span class="caption-text">Thresholds of exponent [Barabasi 2002]</span><a class="headerlink" href="#graph-random-transitions" title="Permalink to this image">¶</a></p>
</div>
<p>In term of degree distribution, w.h.p.</p>
<div class="math notranslate nohighlight">
\[
(1-\varepsilon) \frac{c^{d} e^{-c}}{d !} \leq f_{d}(G) \leq(1+\varepsilon) \frac{c^{d} e^{-c}}{d !}
\]</div>
<p>That is, for large <span class="math notranslate nohighlight">\(N_v\)</span>, <span class="math notranslate nohighlight">\(G\)</span> will have a degree distribution that is like a <strong>Poisson</strong> distribution with mean <span class="math notranslate nohighlight">\(c = p N_v\)</span>. This is intuitive since from the perspective of a vertex <span class="math notranslate nohighlight">\(i \in V\)</span>, it has edge <span class="math notranslate nohighlight">\((i, j)\)</span> w.p. <span class="math notranslate nohighlight">\(p\)</span> for <span class="math notranslate nohighlight">\(N_v - 1\)</span> number of <span class="math notranslate nohighlight">\(j\)</span>, hence its expected degree is <span class="math notranslate nohighlight">\(p(N_v - 1)\)</span>.</p>
<p>Thus, we observe</p>
<ul class="simple">
<li><p><strong>concentrated degree distribution</strong> with exponentially decay tails, rather than broad degree distribution observed in many large-scale real-world networks.</p></li>
<li><p><strong>low clustering</strong>: recall that assortativity is the probability that two neighbors of a randomly chosen vertex are linked is just <span class="math notranslate nohighlight">\(p\)</span>, which tend to zero as <span class="math notranslate nohighlight">\(N_v\)</span> grows.</p></li>
<li><p><strong>small-world property</strong>: the diameter of the graph very like <span class="math notranslate nohighlight">\(\mathcal{O} (\log N_v)\)</span> w.h.p as <span class="math notranslate nohighlight">\(N_v \rightarrow \infty\)</span>.</p></li>
</ul>
<p>Reference</p>
<ul class="simple">
<li><p>Notes on random graphs <a class="reference external" href="https://www.cs.cmu.edu/~avrim/598/chap4only.pdf">CMU</a>, <a class="reference external" href="http://tuvalu.santafe.edu/~aaronc/courses/5352/csci5352_2017_L3.pdf">Sante Fe</a>, <a class="reference external" href="https://www.youtube.com/watch?v=AmZ_MOQ-XwA&amp;ab_channel=LeonidZhukov">HSE video</a>, <a class="reference external" href="http://web.stanford.edu/class/cs224w/slides/14-traditional-generation.pdf">Stanford</a></p></li>
</ul>
</div>
</div>
<div class="section" id="generalized-random-graph-models">
<h3>Generalized Random Graph Models<a class="headerlink" href="#generalized-random-graph-models" title="Permalink to this headline">¶</a></h3>
<p>Equal probability on all graphs of a given order and some particular characteristic(s) <span class="math notranslate nohighlight">\(\eta^*\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathcal{G} (N_v, \eta^*) = \left\{ G = (V, E): \left\vert V \right\vert = N_v, \eta(G) = \eta^*\right\}\]</div>
<p>Erdos Renyi random graph is a particular case of this, with <span class="math notranslate nohighlight">\(\eta^* = N_e\)</span>. <span class="math notranslate nohighlight">\(\eta^*\)</span> can be more general, for instance, degree sequence <span class="math notranslate nohighlight">\(\left\{d_{(1)}, \ldots, d_{\left(N_{v}\right)}\right\}\)</span> in ordered form. Note that since <span class="math notranslate nohighlight">\(N_v\)</span> and <span class="math notranslate nohighlight">\(\bar{d}\)</span> is fixed, due to <span class="math notranslate nohighlight">\(\bar{d} = \frac{2N_e}{N_v}\)</span>, then <span class="math notranslate nohighlight">\(N_e\)</span> is also fixed. Hence, they form a subset of <span class="math notranslate nohighlight">\(\mathcal{G} (N_v, N_e)\)</span>.</p>
<p>Some results</p>
<ul class="simple">
<li><p>suppose <span class="math notranslate nohighlight">\(\eta\)</span> is the first two moments of the degree distribution, under what condition will there be a giant component? [SAND 282, 283]</p></li>
<li><p>suppose <span class="math notranslate nohighlight">\(\left\{ f_d \right\}\)</span> has a power-law form <span class="math notranslate nohighlight">\(f_d = C d^{-\alpha}\)</span>,</p>
<ul>
<li><p>under what condition will there be a giant component? [SAND 5]</p></li>
<li><p>if <span class="math notranslate nohighlight">\(\alpha \in (2,3)\)</span>, the diameter is <span class="math notranslate nohighlight">\(\mathcal{O} (\log N_v)\)</span> and average distance <span class="math notranslate nohighlight">\(\mathcal{O} (\log \log N_v)\)</span> w.h.p. under mild conditions [SAND 87]</p></li>
<li><p>if <span class="math notranslate nohighlight">\(\alpha \in (\frac{7}{3}, 3 )\)</span>, assortativity is <span class="math notranslate nohighlight">\(\mathcal{O} (N_v ^{- \beta})\)</span> where <span class="math notranslate nohighlight">\(\beta = \frac{3\alpha - 7}{\alpha - 1}\)</span>, i.e. the rate is slower than <span class="math notranslate nohighlight">\(N_v ^{-1}\)</span> [SAND 296.IV.B].</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="simulation">
<h3>Simulation<a class="headerlink" href="#simulation" title="Permalink to this headline">¶</a></h3>
<div class="section" id="classical-rgm">
<h4>Classical RGM<a class="headerlink" href="#classical-rgm" title="Permalink to this headline">¶</a></h4>
<p>For some models it is actually possible to produce samples in linear time; for others, it appears that Markov chain Monte Carlo (MCMC) methods are the only realistic alternative.</p>
<ul>
<li><p><span class="math notranslate nohighlight">\(\mathcal{G} (N_v, p)\)</span></p>
<p>A trivial solution is to store <span class="math notranslate nohighlight">\(\binom{N_v}{2} = \mathcal{O} (N_v^2)\)</span> independent Bernoulli random variables, each with success probability <span class="math notranslate nohighlight">\(p\)</span>. When <span class="math notranslate nohighlight">\(p\)</span> is small, majority of these variables will be <span class="math notranslate nohighlight">\(0\)</span>, hence <span class="math notranslate nohighlight">\(\mathcal{O} (N_v^2)\)</span> seems a waste. Can we do better? Hint: for a given vertex <span class="math notranslate nohighlight">\(i\)</span>, consider a sequence of its <span class="math notranslate nohighlight">\(N_v-1\)</span> neighbors <span class="math notranslate nohighlight">\(j\)</span>, such that <span class="math notranslate nohighlight">\(a_{ij} \sim \operatorname{Ber}(p)\)</span>, what’s the expected number of 0’s between two 1’s?</p>
</li>
<li><p><span class="math notranslate nohighlight">\(\mathcal{G} (N_v, N_e)\)</span></p>
<p>It is more cumbersome to use the skipping trick above since edges are correlated: <span class="math notranslate nohighlight">\(\sum_{i\ne j=1}^n a_{ij} = N_e\)</span>. We simply draw <span class="math notranslate nohighlight">\(N_e\)</span> number of distinct pairs from <span class="math notranslate nohighlight">\((i, j) \in V^{(2)}, i\ne j\)</span>, which is a variant of coupon collector’s problem with stopping criteria of reaching <span class="math notranslate nohighlight">\(N_e \le \binom{N_v}{2}\)</span>. This running time is <span class="math notranslate nohighlight">\(\mathcal{O} (N_v + N_e)\)</span> in expectation.</p>
</li>
</ul>
<p>See <a class="reference external" href="http://www.cosinproject.eu/publications/batagelj-pre71-2005.pdf">Batagelj and Brandes</a>.</p>
</div>
<div class="section" id="generalized-rgm">
<h4>Generalized RGM<a class="headerlink" href="#generalized-rgm" title="Permalink to this headline">¶</a></h4>
<p>Sampling GRGM is more challenging since there are more constraints. We focus our discussion upon the case that the degree sequence <span class="math notranslate nohighlight">\(D = \left\{d_{(1)}, \ldots, d_{\left(N_{v}\right)}\right\}\)</span> is be fixed. We introduce two algorithms with input: <span class="math notranslate nohighlight">\(V, D\)</span> and output: <span class="math notranslate nohighlight">\(E\)</span>.</p>
<hr class="docutils" />
<p><strong>Matching Algorithm</strong></p>
<hr class="docutils" />
<ul>
<li><p>create a list containing <span class="math notranslate nohighlight">\(d_{(i)}\)</span> copies of <span class="math notranslate nohighlight">\(v_{(1)}\)</span></p>
<div class="math notranslate nohighlight">
\[
  L = \{ \underbrace{v_{(1)}, \ldots, v_{(1)}}_{d_{(1)} \text{ copies} }, v_{(2)}, \ldots, v_{(N_v)} \}
  \]</div>
</li>
<li><p>randomly choose pairs of elements from <span class="math notranslate nohighlight">\(L\)</span> into <span class="math notranslate nohighlight">\(E\)</span>, removing each pair from <span class="math notranslate nohighlight">\(L\)</span> once chosen.</p></li>
<li><p>return <span class="math notranslate nohighlight">\(E\)</span></p></li>
</ul>
<hr class="docutils" />
<p>Obviously, there can be are multi-edges or loops in <span class="math notranslate nohighlight">\(E\)</span>, hence the corresponding graph is a multi-graph. If that’s the case, just <strong>discard that graph</strong> and then repeat. Under appropriate conditions on the degree sequence, it can be argued that this algorithm will generate graphs from <span class="math notranslate nohighlight">\(\mathcal{G}\)</span> with equal probability. See [SAND 282].</p>
<p>However, when the degree distribution is skewed, e.g. <span class="math notranslate nohighlight">\(d_{(1)}\)</span> is large, it is quite likely to obtain repeated pairs <span class="math notranslate nohighlight">\((v_{(1)}, v_{(j)})\)</span> or <span class="math notranslate nohighlight">\((v_{(1)}, v_{(1)})\)</span>. A solution is to monitor the pairs of vertices being selected and, if a candidate pair matches one in <span class="math notranslate nohighlight">\(E\)</span>, it is <strong>rejected</strong> and another candidate pair is selected instead. This modification will introduce <strong>bias</strong> into the sampling, and the graphs <span class="math notranslate nohighlight">\(G\)</span> thus generated will no longer correspond to a strictly uniform sampling.</p>
<p>Alternatively, we can instead sample so as to avoid repeating existing matches in the first place. See [SAND 81] that developed for uniformly sampling <span class="math notranslate nohighlight">\(r \times c\)</span> matrices <span class="math notranslate nohighlight">\(\boldsymbol{M}\)</span> of non-negative integers with fixed marginal totals.</p>
<hr class="docutils" />
<p><strong>Switching Algorithm</strong> (Aka rewiring algorithms)</p>
<hr class="docutils" />
<ul class="simple">
<li><p>begin with a graph that has the prescribed degree sequence</p></li>
<li><p>modify the connectivity of that graph through a succession of simple changes named ‘switching’: a pair of edges in the current graph <span class="math notranslate nohighlight">\(e_1 = (u_1, v_1)\)</span> and <span class="math notranslate nohighlight">\(e_2 = (u_2, v_2)\)</span> are randomly selected and replaced by the new edges <span class="math notranslate nohighlight">\(\left\{ u_1, v_2 \right\}\)</span> and <span class="math notranslate nohighlight">\(\left\{ u_2, v_1 \right\}\)</span>. If either of the latter already exists, then the proposed switch is abandoned.</p></li>
</ul>
<hr class="docutils" />
<p>It falls within the realm of MCMC methods. In practice, it is typical to let the algorithm run for some time before beginning to collect sample graphs <span class="math notranslate nohighlight">\(G\)</span>. There is currently no  theory to indicate just how long of a preliminary period is necessary. Milo et al. [SAND 279] cite empirical evidence to suggest a factor of <span class="math notranslate nohighlight">\(100 N_e\)</span> can be more than sufficient.</p>
<p>To ensure that the algorithm asymptotically yields strictly uniform sampling from <span class="math notranslate nohighlight">\(\mathcal{G}\)</span>, there are some certain formal conditions. See [SAND 322].</p>
<p>MCMC can be used to generate GRG uniformly from other types of collections <span class="math notranslate nohighlight">\(\mathcal{G}\)</span> with additional characteristics beyond the degree sequence. However, that development of the corresponding theory, verifying the assumptions underlying Markov chain <strong>convergence</strong>, currently appears to lag far behind the pace of algorithm development.</p>
</div>
</div>
<div class="section" id="application">
<h3>Application<a class="headerlink" href="#application" title="Permalink to this headline">¶</a></h3>
<div class="section" id="hidden-population-size">
<h4>Hidden Population Size<a class="headerlink" href="#hidden-population-size" title="Permalink to this headline">¶</a></h4>
<p>In previous section we derived a design-based <a class="reference internal" href="13-sampling-and-estimation.html#sampling-hidden-pop-size"><span class="std std-ref">estimator</span></a> of hidden population size. Here, we describe a model-based estimator using random graphs.</p>
<p>Assume that the hidden population graph <span class="math notranslate nohighlight">\(G = (V, E)\)</span> is from a collection <span class="math notranslate nohighlight">\(\mathcal{G} (N_v, p)\)</span> of random graphs.</p>
<p>Note that now, <span class="math notranslate nohighlight">\(A_{ij}\)</span> is also a random variables, i.e. <span class="math notranslate nohighlight">\(A_{ij} = \mathbb{I} \left\{ (i, j) \in E  \right\}\)</span> with success probability <span class="math notranslate nohighlight">\(p\)</span>. Moreover, <span class="math notranslate nohighlight">\(A_{ij}\)</span> are independent with <span class="math notranslate nohighlight">\(Z_i\)</span>. Hence, the moments are</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{aligned}
\mathbb{E}\left(N_{v}^{*}\right) &amp;=\mathbb{E}\left(\sum_{i} Z_{i}\right)=N_{v} p_{0} \\
\mathbb{E}\left(M_{1}\right) &amp;=\mathbb{E}\left(\sum_{i \neq j} Z_{i} Z_{j} A_{i j}\right)= p_{0}^{2} \cdot N_v\left(N_{v}-1\right)p \\
\mathbb{E}\left(M_{2}\right) &amp;=\mathbb{E}\left(\sum_{i \neq j} Z_{i}\left(1-Z_{j}\right) A_{i j}\right)= p_{0}\left(1-p_{0}\right) \cdot N_v\left(N_{v}-1\right)p
\end{aligned}
\end{split}\]</div>
<p>After setting RHS equal to the observed values <span class="math notranslate nohighlight">\(n, m_1, m_2\)</span> of LHS, we have</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{aligned}
\hat{p}_{0} &amp;=m_{1} /\left(m_{1}+m_{2}\right) \\
\hat{N}_{v} &amp;=n\left(m_{1}+m_{2}\right) / m_{1}
\hat{p} &amp;=m_{1}\left(m_{1}+m_{2}\right) / n\left[(n-1) m_{1}+n m_{2}\right] \\
\end{aligned}
\end{split}\]</div>
<p>Note that the estimates of <span class="math notranslate nohighlight">\(p_0\)</span> and <span class="math notranslate nohighlight">\(N_v\)</span> are the same as those in the design-based method. There is another method using maximum conditional likelihood, see [SAND 154].</p>
</div>
<div class="section" id="assessing-significance">
<h4>Assessing Significance<a class="headerlink" href="#assessing-significance" title="Permalink to this headline">¶</a></h4>
<p>As described, given <span class="math notranslate nohighlight">\(\eta(G^{obs})\)</span>, we want to find how in some sense unusual or unexpected it is. An important issue is determining which <span class="math notranslate nohighlight">\(\mathcal{G}\)</span> to use as reference. For instance, to assess the significance of the number of distinct triangles in <span class="math notranslate nohighlight">\(G^{obs}\)</span>, a reasonable reference <span class="math notranslate nohighlight">\(\mathcal{G}\)</span> should have the same number of edges as that of <span class="math notranslate nohighlight">\(G^{obs}\)</span>. In practice,</p>
<ul class="simple">
<li><p>it is common to control for the degree sequence observed in <span class="math notranslate nohighlight">\(G^{obs}\)</span>.</p></li>
<li><p>sometimes also control other factors: if we knew that certain groups were present within the vertex set <span class="math notranslate nohighlight">\(V^{obs}\)</span>, we might wish to maintain the number of edges between and within each group.</p></li>
</ul>
<p>For example, for an observed graph <span class="math notranslate nohighlight">\(G^{obs}\)</span>, we find that <span class="math notranslate nohighlight">\(\operatorname{clus}_T(G^{obs}) = 0.2257\)</span>. A reference collection <span class="math notranslate nohighlight">\(\mathcal{G}\)</span> can be</p>
<ol class="simple">
<li><p><span class="math notranslate nohighlight">\(\mathcal{G} (N_v, N_e)\)</span>, or</p></li>
<li><p><span class="math notranslate nohighlight">\(\mathcal{G} (N_v, N_e, d)\)</span>, where <span class="math notranslate nohighlight">\(d\)</span> is the degree distribution.</p></li>
</ol>
<p>Note <span class="math notranslate nohighlight">\(\left\vert \mathcal{G} (N_v, N_e) \right\vert = \binom{\binom{N_v}{2} }{N_e}\)</span> is quite large, and <span class="math notranslate nohighlight">\(\left\vert \mathcal{G} (N_v, N_e, f_d) \right\vert\)</span> is much smaller than the former but also large, we use MCMC to simulate the uniform sampling of <span class="math notranslate nohighlight">\(10,000\)</span> random graphs <span class="math notranslate nohighlight">\(G\)</span> from <span class="math notranslate nohighlight">\(\mathcal{G}\)</span>, and compute <span class="math notranslate nohighlight">\(\eta(G) = \operatorname{clus}_T(G)\)</span></p>
<ul class="simple">
<li><p>in the first case, when the number of edges was fixed, only 3 of our 10,000 samples resulted in a graph with higher clustering coefficient,</p></li>
<li><p>whereas in the second case, when the degree distribution was fixed, all of the samples resulted in lower clustering coefficient.</p></li>
</ul>
<p>We therefore have strong evidence to reject the hypothesis that the network can be viewed as a uniform sample under either random graph model. As a result, we conclude that the network graph shows <strong>markedly</strong> greater transitivity than random graphs of comparable magnitude (i.e., with respect to order and size) or connectivity (i.e., with respect to degree distribution).</p>
<div class="figure align-default" id="graph-karate">
<a class="reference internal image-reference" href="../_images/graph-karate.png"><img alt="" src="../_images/graph-karate.png" style="width: 50%;" /></a>
<p class="caption"><span class="caption-number">Fig. 194 </span><span class="caption-text">Observed graph of Zachary’s ‘karate club [Kolaczyk 2009]</span><a class="headerlink" href="#graph-karate" title="Permalink to this image">¶</a></p>
</div>
<div class="figure align-default" id="rgm-assess-clustering">
<a class="reference internal image-reference" href="../_images/rgm-assess-clustering.png"><img alt="" src="../_images/rgm-assess-clustering.png" style="width: 80%;" /></a>
<p class="caption"><span class="caption-number">Fig. 195 </span><span class="caption-text">Histograms of simulated clustering coefficients with fixed <span class="math notranslate nohighlight">\(N_v, N_e\)</span> (left) and <span class="math notranslate nohighlight">\(f_d\)</span> (right).</span><a class="headerlink" href="#rgm-assess-clustering" title="Permalink to this image">¶</a></p>
</div>
<p>It’s also worth observing that the right distribution is bimodal.</p>
<ul class="simple">
<li><p>those in in the right-hand mode tended to often be characterized by two large and somewhat distinct clusters, as in the original network.</p></li>
<li><p>those in the left appeared to have more diffuse clusters or even just one large cluster.</p></li>
</ul>
<p>Meanwhile, for the <span class="math notranslate nohighlight">\(\mathcal{G} (N_v, N_e, d)\)</span> case, due to conditioning on degree, coupled with the invariance of <span class="math notranslate nohighlight">\(\operatorname{clus} _T\)</span> under isomorphism, the effective size of the sample space becomes quite small. In the 10,000 trials run, there were only 25 different values of <span class="math notranslate nohighlight">\(\operatorname{clus} _T\)</span>, and 17 of them takes 99% of the mass.</p>
</div>
<div class="section" id="detecting-network-motifs">
<h4>Detecting Network Motifs<a class="headerlink" href="#detecting-network-motifs" title="Permalink to this headline">¶</a></h4>
<dl class="simple myst">
<dt>Definition (motif)</dt><dd><p>Motif defined by [SAND 218, 278] are small subgraphs occurring <strong>far more frequently</strong> in a given network than in comparable random graphs.</p>
</dd>
</dl>
<p>Motivation: many large, complex networks may perhaps be constructed (at least in part) of smaller, comparatively simple ‘building blocks’. Network motif detection seeks to identify possible subgraph configurations of this nature.</p>
<p>Let</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\mathcal{G} _k =\left\{ g: \left\vert V(g) \right\vert= k  \right\}\)</span> be a collection of all possible non-isomorphic <span class="math notranslate nohighlight">\(k\)</span>-vertex connected subgraphs.</p></li>
<li><p><span class="math notranslate nohighlight">\(L_k = \left\vert \mathcal{G} _k \right\vert\)</span> be the cardinality</p></li>
<li><p><span class="math notranslate nohighlight">\(N_i\)</span> be the number of occurrences of the <span class="math notranslate nohighlight">\(i\)</span>-th element <span class="math notranslate nohighlight">\(g_i \in \mathcal{G} _k\)</span> in <span class="math notranslate nohighlight">\(G\)</span>, for <span class="math notranslate nohighlight">\(i = 1, 2, \ldots, L_k\)</span></p></li>
</ul>
<p>For instance, in directed graphs, when <span class="math notranslate nohighlight">\(k=3\)</span>, <span class="math notranslate nohighlight">\(\mathcal{G}_3\)</span> is the collection of 13 connected subgraphs:</p>
<div class="figure align-default" id="graph-motif-k3">
<a class="reference internal image-reference" href="../_images/graph-motif-k3.png"><img alt="" src="../_images/graph-motif-k3.png" style="width: 70%;" /></a>
<p class="caption"><span class="caption-number">Fig. 196 </span><span class="caption-text">Connected subgraphs when <span class="math notranslate nohighlight">\(k=3\)</span> in directed graphs [<a class="reference external" href="http://pbil.univ-lyon1.fr/members/sagot/htdocs/coursesENS/motifs.pdf">pic</a>]</span><a class="headerlink" href="#graph-motif-k3" title="Permalink to this image">¶</a></p>
</div>
<p>Define a proportion as our <span class="math notranslate nohighlight">\(\eta(G)\)</span>:</p>
<div class="math notranslate nohighlight">
\[
F_i = \frac{N_i}{\sum_{i^\prime =1}^{L_k} N_{i ^\prime}}
\]</div>
<p>Then, analogous to what was described in the previous section, each value <span class="math notranslate nohighlight">\(F_i\)</span> is compared to an appropriate reference distribution <span class="math notranslate nohighlight">\(\mathbb{P}_{\mathcal{G} } (i)\)</span>. Subgraphs for whom the value <span class="math notranslate nohighlight">\(F_i\)</span> is found to be extreme are declared to be network motifs for building <span class="math notranslate nohighlight">\(G^{obs}\)</span>.</p>
<p>Note that we for a given choice of <span class="math notranslate nohighlight">\(k\)</span>, we need to count the number <span class="math notranslate nohighlight">\(N_i\)</span> for <span class="math notranslate nohighlight">\(i = 1,\ldots, L_k\)</span>, but <span class="math notranslate nohighlight">\(L_k\)</span> grows quite large with <span class="math notranslate nohighlight">\(k\)</span>. As shown above, in directed graphs, <span class="math notranslate nohighlight">\(L_3=13\)</span>, and then <span class="math notranslate nohighlight">\(L_4 = 199, L_5 = 9364\)</span>. To overcome this, some sampling techniques can be used. Specifically, if <span class="math notranslate nohighlight">\(k\)</span>-vertex subgraphs <span class="math notranslate nohighlight">\(H\)</span> are sampled in some fashion, then an unbiased estimate of the total number <span class="math notranslate nohighlight">\(N_i\)</span> of a given subgraph type is just</p>
<div class="math notranslate nohighlight">
\[
\widehat{N}_i = \sum_{H\text{ of type }i } \pi_H^{-1}
\]</div>
<p>where <span class="math notranslate nohighlight">\(\pi_H\)</span> is the inclusion probability for <span class="math notranslate nohighlight">\(H\)</span>. Natural (although biased) estimates <span class="math notranslate nohighlight">\(\hat{F}_i\)</span> of the corresponding relative frequencies <span class="math notranslate nohighlight">\(F_i\)</span> are obtained through direct substitution of <span class="math notranslate nohighlight">\(\hat{N}_i\)</span> to the previous equation.</p>
<p>For other sampling method, see SAND pg.168.</p>
</div>
</div>
</div>
<div class="section" id="real-world-models">
<h2>Real-world Models<a class="headerlink" href="#real-world-models" title="Permalink to this headline">¶</a></h2>
<p>Real-world graphs do not like the <span class="math notranslate nohighlight">\(G_{n,p}\)</span> random graphs. Many real networks have high clustering.</p>
<div class="figure align-default" id="graph-real-vs-random">
<a class="reference internal image-reference" href="../_images/graph-real-vs-random.png"><img alt="" src="../_images/graph-real-vs-random.png" style="width: 70%;" /></a>
<p class="caption"><span class="caption-number">Fig. 197 </span><span class="caption-text">Comparison of real-world graphs and random graphs with the same average degree [Leskovec 2021]</span><a class="headerlink" href="#graph-real-vs-random" title="Permalink to this image">¶</a></p>
</div>
<p>We introduce two random graph models to mimic the high clustering, high diameter properties in real-world graphs.</p>
<p>Note all these models (including <span class="math notranslate nohighlight">\(G_{n, p}\)</span>) impose prior assumption of the graph generation processes. Deep graph generative models, however, learn the graph generation process from raw data.</p>
<div class="section" id="small-world-models">
<h3>Small-World Models<a class="headerlink" href="#small-world-models" title="Permalink to this headline">¶</a></h3>
<p>Developed by Watts and Strogatz that mimic certain observed ‘real-world’ properties: <strong>high</strong> levels of clustering, but <strong>small</strong> distances between most nodes.</p>
<p>To model this, consider starting with a lattice, and randomly rewiring a a few edges. Specifically,</p>
<ol class="simple">
<li><p>begin with a set of <span class="math notranslate nohighlight">\(N_v\)</span> vertices, arranged in a ‘ring’-like fashion</p></li>
<li><p>join each vertex to a <span class="math notranslate nohighlight">\(r\)</span> of its neighbors to each side.</p></li>
<li><p>for each edge, w.p. <span class="math notranslate nohighlight">\(p\)</span>, independently move one of its end to be incident to another uniformly randomly chosen vertex, but avoid loops and multi-edges.</p></li>
</ol>
<div class="figure align-default" id="graph-ws-lattice">
<a class="reference internal image-reference" href="../_images/graph-WS-lattice.png"><img alt="" src="../_images/graph-WS-lattice.png" style="width: 30%;" /></a>
<p class="caption"><span class="caption-number">Fig. 198 </span><span class="caption-text">Example of a Watts-Strogatz ‘small-world’ network graph. [Kolaczyk 2009]</span><a class="headerlink" href="#graph-ws-lattice" title="Permalink to this image">¶</a></p>
</div>
<p>For a lattice <span class="math notranslate nohighlight">\(G\)</span> with parameter <span class="math notranslate nohighlight">\(r\)</span> described above, it is easy to find</p>
<ul class="simple">
<li><p>high level of clustering: <span class="math notranslate nohighlight">\(\operatorname{clus}_T (G) = \frac{3r-3}{4r-2} \approx \frac{3}{4}\)</span></p></li>
<li><p>long diameter: <span class="math notranslate nohighlight">\(\operatorname{diam}(G) = \frac{N_v}{2r}\)</span></p></li>
<li><p>long average distance: <span class="math notranslate nohighlight">\(\bar{l} = \frac{N_v}{4r}\)</span></p></li>
</ul>
<p>But, addition of a few randomly rewired edges has the effect of producing ‘short-cuts’ in the graph, but it takes a lot of randomness to ruin clustering. In numerical simulation shown below, after rewiring with some small <span class="math notranslate nohighlight">\(p\)</span>, we have <span class="math notranslate nohighlight">\(\bar{l} = \mathcal{O} (\log N_v)\)</span> while keep clustering coefficient close to <span class="math notranslate nohighlight">\(\frac{3}{4}\)</span>.</p>
<div class="figure align-default" id="graph-ws-simulation">
<a class="reference internal image-reference" href="../_images/graph-WS-simulation.png"><img alt="" src="../_images/graph-WS-simulation.png" style="width: 30%;" /></a>
<p class="caption"><span class="caption-number">Fig. 199 </span><span class="caption-text">Simulation results of clustering coefficient <span class="math notranslate nohighlight">\(\operatorname{clus}(G)\)</span> (solid) and <span class="math notranslate nohighlight">\(\bar{l}\)</span> (dashed) as a function of <span class="math notranslate nohighlight">\(p\)</span>, both have been
normalized by their largest values</span><a class="headerlink" href="#graph-ws-simulation" title="Permalink to this image">¶</a></p>
</div>
<p>Essentially, it interpolate between regular lattice graphs and <span class="math notranslate nohighlight">\(G_{n,p}\)</span> random graph.</p>
<div class="figure align-default" id="graph-sw-interpolate">
<a class="reference internal image-reference" href="../_images/graph-sw-interpolate.png"><img alt="" src="../_images/graph-sw-interpolate.png" style="width: 70%;" /></a>
<p class="caption"><span class="caption-number">Fig. 200 </span><span class="caption-text">Small world model as interpolation [Leskovec 2021]</span><a class="headerlink" href="#graph-sw-interpolate" title="Permalink to this image">¶</a></p>
</div>
<p>However, the closed form for <span class="math notranslate nohighlight">\(\operatorname{clus}(G)\)</span> and <span class="math notranslate nohighlight">\(\bar{l}\)</span> are still open problems.</p>
<p>Variation:</p>
<ul class="simple">
<li><p>both ends of an edge are rewired [SAND 23]</p></li>
<li><p>no edges are rewired, but some small number of new edges are added to randomly selected pairs of vertices [SAND 284, 298]</p></li>
<li><p>add edge <span class="math notranslate nohighlight">\((u, v)\)</span> w.p. inversely proportional to <span class="math notranslate nohighlight">\(\operatorname{dist} (u,v)\)</span>, i.e. <span class="math notranslate nohighlight">\(p \propto (\operatorname{dist} )^{-r}\)</span> for some <span class="math notranslate nohighlight">\(r &gt; 0\)</span>. [SAND 229, 231].</p></li>
</ul>
</div>
<div class="section" id="kronecker-graph-model">
<h3>Kronecker Graph Model<a class="headerlink" href="#kronecker-graph-model" title="Permalink to this headline">¶</a></h3>
<p>Kronecker product <span class="math notranslate nohighlight">\(\otimes\)</span> is a way of generating self-similar matrices. It is defined as</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\mathbf{C}=\mathbf{A} \otimes \mathbf{B} \doteq\left(\begin{array}{cccc}
a_{1,1} \mathbf{B} &amp; a_{1,2} \mathbf{B} &amp; \ldots &amp; a_{1, m} \mathbf{B} \\
a_{2,1} \mathbf{B} &amp; a_{2,2} \mathbf{B} &amp; \ldots &amp; a_{2, m} \mathbf{B} \\
\vdots &amp; \vdots &amp; \ddots &amp; \vdots \\
a_{n, 1} \mathbf{B} &amp; a_{n, 2} \mathbf{B} &amp; \ldots &amp; a_{n, m} \mathbf{B}
\end{array}\right)
\end{split}\]</div>
<p>If <span class="math notranslate nohighlight">\(\boldsymbol{A}, \boldsymbol{B}\)</span> are the adjacency matrix of a graph, then <strong>Kronecker power</strong> mimics <strong>recursive</strong> graph/community growth.</p>
<div class="figure align-default" id="graph-kronecker-self">
<a class="reference internal image-reference" href="../_images/graph-kronecker-self.png"><img alt="" src="../_images/graph-kronecker-self.png" style="width: 70%;" /></a>
<p class="caption"><span class="caption-number">Fig. 201 </span><span class="caption-text">Kronecker power of a adjacency matrix [Leskovec 2021]</span><a class="headerlink" href="#graph-kronecker-self" title="Permalink to this image">¶</a></p>
</div>
<p>Kronecker graph model uses it to construct a random matrix with Bernoulli matrix.</p>
<ol class="simple">
<li><p>Create <span class="math notranslate nohighlight">\(N_1 \times N_1\)</span> probability matrix <span class="math notranslate nohighlight">\(\boldsymbol{\Theta}_1\)</span>, where all entry values are in <span class="math notranslate nohighlight">\((0, 1)\)</span>.</p></li>
<li><p>Compute the <span class="math notranslate nohighlight">\(k\)</span>-th Kronecker power <span class="math notranslate nohighlight">\(\boldsymbol{\Theta}_k\)</span>.</p></li>
<li><p>For each entry <span class="math notranslate nohighlight">\(p_{uv}\)</span> in matrix <span class="math notranslate nohighlight">\(\boldsymbol{\Theta}_k\)</span>, include an edge <span class="math notranslate nohighlight">\((u, v)\)</span> with probability <span class="math notranslate nohighlight">\(p_{uv}\)</span>.</p></li>
</ol>
<div class="figure align-default" id="graph-kronecker-prob">
<a class="reference internal image-reference" href="../_images/graph-kronecker-prob.png"><img alt="" src="../_images/graph-kronecker-prob.png" style="width: 80%;" /></a>
<p class="caption"><span class="caption-number">Fig. 202 </span><span class="caption-text">Adjacency matrix in Kronecker graph models [Leskovec 2021]</span><a class="headerlink" href="#graph-kronecker-prob" title="Permalink to this image">¶</a></p>
</div>
<p>Obviously, there involves <span class="math notranslate nohighlight">\(\mathcal{O} (N_v^2)\)</span> Bernoulli random variables. To improve this, we can exploit the recursive structure of Kronecker graphs.</p>
<p>It is quite close to real graphs.</p>
<div class="figure align-default" id="graph-kronecker-vs-real">
<a class="reference internal image-reference" href="../_images/graph-kronecker-vs-real.png"><img alt="" src="../_images/graph-kronecker-vs-real.png" style="width: 70%;" /></a>
<p class="caption"><span class="caption-number">Fig. 203 </span><span class="caption-text">A Kronecker graph vs a real graph [Leskovec 2021]</span><a class="headerlink" href="#graph-kronecker-vs-real" title="Permalink to this image">¶</a></p>
</div>
</div>
</div>
<div class="section" id="growth-models">
<h2>Growth Models<a class="headerlink" href="#growth-models" title="Permalink to this headline">¶</a></h2>
<p>Many networks grow or otherwise evolve in time, e.g. WWW and citation networks. Problems of interest include</p>
<ul class="simple">
<li><p>how the graph changes? how to model?</p>
<ul>
<li><p>vertex preference, fitness, copying, etc</p></li>
</ul>
</li>
<li><p>as <span class="math notranslate nohighlight">\(t\)</span> goes large, some properties emerge</p></li>
</ul>
<div class="section" id="preferential-attachment">
<h3>Preferential Attachment<a class="headerlink" href="#preferential-attachment" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>principle: the rich get richer.</p>
<ul>
<li><p>observe: in WWW, often web pages to which many other pages point will tend to accumulate increasingly greater numbers of links as time goes on.</p></li>
</ul>
</li>
<li><p>want to mimic: broad degree distributions, as observed in many large, real-world networks</p></li>
</ul>
<p>We introduce <strong>Barabasi-Albert Model</strong>, which view degree as ‘richness’.</p>
<ul class="simple">
<li><p>Start with an initial graph <span class="math notranslate nohighlight">\(G{(0)}\)</span> of <span class="math notranslate nohighlight">\(N(0)_v\)</span> vertices and <span class="math notranslate nohighlight">\(N^{(0)}_e\)</span> edges.</p></li>
<li><p>At stage <span class="math notranslate nohighlight">\(t = 1,2, \ldots\)</span>, the current graph <span class="math notranslate nohighlight">\(G^{(t−1)}\)</span> is modified to create a new graph <span class="math notranslate nohighlight">\(G^{t}\)</span> by</p>
<ul>
<li><p>adding a new vertex of degree <span class="math notranslate nohighlight">\(m\ge 1\)</span>, where the <span class="math notranslate nohighlight">\(m\)</span> new edges are attached to <span class="math notranslate nohighlight">\(m\)</span> different vertices in <span class="math notranslate nohighlight">\(G^{(t−1)}\)</span>, with probability <span class="math notranslate nohighlight">\(\frac{d_v}{\sum_{v ^\prime \in V} d_{v ^\prime}}\)</span> for <span class="math notranslate nohighlight">\(v\)</span> to be connected (preferential to those with <strong>higher</strong> degrees).</p></li>
<li><p><span class="math notranslate nohighlight">\(G^{t}\)</span> will have <span class="math notranslate nohighlight">\(N_v^{(t)} = N_v^{(0)} + t\)</span> vertices and <span class="math notranslate nohighlight">\(N_e^{(t)} = N_e^{(0)} + tm\)</span> edges</p></li>
</ul>
</li>
</ul>
<p>We would expect that a number of vertices of comparatively high degree (“rich”) should gradually emerge as <span class="math notranslate nohighlight">\(t\)</span> increases.</p>
<p>How to select <span class="math notranslate nohighlight">\(m\)</span> vertices exactly? We introduce the <strong>linearized-chord diagram</strong> (LCD) model. For simulation of LCD in linear time, see [SAND 23].</p>
<p>For the case <span class="math notranslate nohighlight">\(m=1\)</span>,</p>
<ul>
<li><p>begin with <span class="math notranslate nohighlight">\(G^{(1)}\)</span> consisting of a single vertex with a loop</p></li>
<li><p>for <span class="math notranslate nohighlight">\(t = 2, 3 \ldots,\)</span></p>
<ul class="simple">
<li><p>add the vertex <span class="math notranslate nohighlight">\(v_t\)</span> to <span class="math notranslate nohighlight">\(G^{(t-1)}\)</span> with an edge to a vertex <span class="math notranslate nohighlight">\(v_s\)</span> for <span class="math notranslate nohighlight">\(1 \le s \le t\)</span> chosen randomly w.p.</p></li>
</ul>
<div class="math notranslate nohighlight">
\[\begin{split}
  \mathbb{P}(s=j)=\left\{\begin{array}{ll}
  d_{G(t-1)}\left(v_{j}\right) /(2 t-1), &amp; \text { if } 1 \leq j \leq t-1 ， \\
  1 /(2 t-1), &amp; \text { if } j=t,
  \end{array}\right.
  \end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(d_{G^{(t-1)}}\left(v_{j}\right)\)</span> is the degree of <span class="math notranslate nohighlight">\(v_j\)</span> at time <span class="math notranslate nohighlight">\(t-1\)</span>.</p>
</li>
</ul>
<p>For the case <span class="math notranslate nohighlight">\(m &gt; 1\)</span>, we repeat the above process for <span class="math notranslate nohighlight">\(m\)</span> steps, after which we contract the added <span class="math notranslate nohighlight">\(m\)</span> vertices into one, with the <span class="math notranslate nohighlight">\(m\)</span> edges retained. Clearly, this formulation allows for loops and multi-edges. However, these should occur relatively rarely, and the precision gained by this formulation is necessary for rigorously deriving mathematical results regarding model properties for <span class="math notranslate nohighlight">\(G^{(t)}\)</span>:</p>
<ul>
<li><p><strong>connected w.h.p.</strong> (not connected if, e.g., <span class="math notranslate nohighlight">\(m=1, j=t\)</span> self-loop).</p></li>
<li><p><strong>power-law degree w.h.p.</strong> <span class="math notranslate nohighlight">\(t\)</span> tends to infinity, <span class="math notranslate nohighlight">\(G^{(t)}\)</span> have degree distributions that tend to a power-law form <span class="math notranslate nohighlight">\(d^{-\alpha}\)</span> with <span class="math notranslate nohighlight">\(\alpha = 3\)</span>. It can be shown that w.h.p. for any <span class="math notranslate nohighlight">\(\epsilon\)</span> and every <span class="math notranslate nohighlight">\(0 \le d\le \left(N_{v}^{(t)}\right)^{1 / 5}\)</span>,</p>
<div class="math notranslate nohighlight">
\[
  f_{d}\left(G^{(t)}\right) \in (1 \pm \varepsilon) f_{d, m} \quad \text{where } f_{d, m}= \frac{2m(m+1)}{(d+2)(d+1) d}
  \]</div>
<p>which behaves like <span class="math notranslate nohighlight">\(d^{-3}\)</span> for <span class="math notranslate nohighlight">\(d\)</span> large relative to <span class="math notranslate nohighlight">\(m\)</span>.</p>
</li>
<li><p><strong>small diameter</strong>: when <span class="math notranslate nohighlight">\(m=1\)</span>, diameter <span class="math notranslate nohighlight">\(\operatorname{diam}(G^{(t)}) = \mathcal{O} (\log N_v^{(t)})\)</span>. when <span class="math notranslate nohighlight">\(m&gt;1\)</span>, it is <span class="math notranslate nohighlight">\(\mathcal{O} \left( \frac{\log N_v^{(t)}}{\log \log N_v^{(t)}} \right)\)</span>, which is a bit smaller still.</p></li>
<li><p><strong>less clustering</strong></p>
<div class="math notranslate nohighlight">
\[
  \mathbb{E} [\operatorname{clus}_T (G^{(t)}) ]  = \mathcal{O} \left( \frac{m-1}{8} \frac{\left(\log N_{v}^{(t)}\right)^{2}}{N_{v}^{(t)}} \right)
  \]</div>
<p>which is only a little better than <span class="math notranslate nohighlight">\(N_v^{-1}\)</span> behavior in the case of classical random graph models.</p>
</li>
</ul>
<p>There many extensions and variations, such as</p>
<ul class="simple">
<li><p>clustering, diameter, etc</p></li>
<li><p>consider other ‘fitness’ or inherent quality of vertices as ‘richness’?</p></li>
<li><p>allow <span class="math notranslate nohighlight">\(m\)</span> to vary, or dynamic addition and removal of edges?</p></li>
<li><p>add offset <span class="math notranslate nohighlight">\(d^*\)</span> to <span class="math notranslate nohighlight">\(d_v + d^*\)</span>, or use powers <span class="math notranslate nohighlight">\(d_v^\gamma\)</span>.</p></li>
</ul>
<p>The main concern include</p>
<ul class="simple">
<li><p>whether or not a power-law limiting distribution is achieved?</p></li>
<li><p>if so, how is <span class="math notranslate nohighlight">\(\alpha\)</span> related to model parameters?</p></li>
</ul>
<p>See [SAND 6, 296, 41].</p>
</div>
<div class="section" id="copying-models">
<h3>Copying Models<a class="headerlink" href="#copying-models" title="Permalink to this headline">¶</a></h3>
<p>Copying model is distinct from preferential attachment but can also produce power-law degree distributions.</p>
<div class="margin sidebar">
<p class="sidebar-title"></p>
<p>When <span class="math notranslate nohighlight">\(p=1\)</span>, each new vertex <span class="math notranslate nohighlight">\(v\)</span> is connected to <span class="math notranslate nohighlight">\(G ^{(t-1)}\)</span> by fully copying the edges of the randomly chosen vertex <span class="math notranslate nohighlight">\(u\)</span>, hence it is named ‘copying’ model.</p>
</div>
<div class="section" id="full-copying">
<h4>Full Copying<a class="headerlink" href="#full-copying" title="Permalink to this headline">¶</a></h4>
<p>Chung, Lu, Dewey, and Galas [SAND 89]</p>
<ul class="simple">
<li><p>beginning with an initial graph <span class="math notranslate nohighlight">\(G^{(0)}\)</span>,</p></li>
<li><p>for <span class="math notranslate nohighlight">\(t = 1, \ldots,\)</span></p>
<ul>
<li><p>choose a vertex <span class="math notranslate nohighlight">\(u\)</span> from graph <span class="math notranslate nohighlight">\(g^{(t)}\)</span></p></li>
<li><p>add a new vertex <span class="math notranslate nohighlight">\(v\)</span>, join it with each of the neighbors of <span class="math notranslate nohighlight">\(u\)</span> independently w.p. <span class="math notranslate nohighlight">\(p\)</span></p></li>
</ul>
</li>
</ul>
<p>The degree distribution <span class="math notranslate nohighlight">\(f_d{G ^{(t)} }\)</span> will tend to a power-law form <span class="math notranslate nohighlight">\(d^{-\alpha}\)</span> w.h.p,, where <span class="math notranslate nohighlight">\(\alpha\)</span> satisfying</p>
<div class="math notranslate nohighlight">
\[
p(\alpha - 1) = 1- p^{\alpha - 1}
\]</div>
<p>This equation will have two solutions <span class="math notranslate nohighlight">\(\alpha\)</span> for any given <span class="math notranslate nohighlight">\(p\)</span>, but only one will be stable.</p>
<ul class="simple">
<li><p>For <span class="math notranslate nohighlight">\(p &gt; 0.5671 \ldots\)</span>, the stable solution is <span class="math notranslate nohighlight">\(\alpha = 1\)</span></p></li>
<li><p>For <span class="math notranslate nohighlight">\(p=1/2\)</span>, it is <span class="math notranslate nohighlight">\(\alpha = 2\)</span></p></li>
<li><p>For <span class="math notranslate nohighlight">\(p=1\)</span>, no power-law behavior. To achieve power-law, it is sufficient to allow partial duplication to occur some fraction <span class="math notranslate nohighlight">\(q \in (0,1)\)</span> of the times.</p></li>
</ul>
<div class="figure align-default" id="graph-copying">
<a class="reference internal image-reference" href="../_images/graph-copying.png"><img alt="" src="../_images/graph-copying.png" style="width: 30%;" /></a>
<p class="caption"><span class="caption-number">Fig. 204 </span><span class="caption-text">Power-law exponent <span class="math notranslate nohighlight">\(\alpha\)</span> as a function of <span class="math notranslate nohighlight">\(p\)</span> in copying model. [Kolaczyk 2009]</span><a class="headerlink" href="#graph-copying" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="partial-copying">
<h4>Partial Copying<a class="headerlink" href="#partial-copying" title="Permalink to this headline">¶</a></h4>
<p>Kleinberg et al. [SAND 232] and by Kumar et al. [SAND 241]</p>
<ul class="simple">
<li><p>for <span class="math notranslate nohighlight">\(i=1, \ldots, m\)</span>, w.p. <span class="math notranslate nohighlight">\(1-\beta\)</span> copying the <span class="math notranslate nohighlight">\(i\)</span>-th neighbor of <span class="math notranslate nohighlight">\(u\)</span>, and w.p. <span class="math notranslate nohighlight">\(\beta\)</span> to form edge between <span class="math notranslate nohighlight">\(v\)</span> and some vertex uniformly selected from <span class="math notranslate nohighlight">\(G^{(t-1)}\)</span>. [SAND 232 241].</p></li>
</ul>
<p>Intuition</p>
<ul class="simple">
<li><p>In WWW, when an individual setting up a new web page,</p>
<ul>
<li><p>On the one hand, the web page is likely being set up in the context of some established topic(s), and thus will probably include a number of established links in that topic.</p></li>
<li><p>On the other hand, an individual also brings their own perspective to the topic, and will be expected to generate some previously unseen links as well</p></li>
</ul>
</li>
<li><p>In protein interaction networks, structure of genes associated with the proteins as typically evolving through duplication, as above, but periodically arising instead through mutation.</p></li>
</ul>
<p>Properties</p>
<ul>
<li><p>The degree distribution <span class="math notranslate nohighlight">\(\left\{ f_d(G ^{(t)} ) \right\}\)</span> also tends to a power law. For each <span class="math notranslate nohighlight">\(d &gt;0\)</span>, as <span class="math notranslate nohighlight">\(t \rightarrow \infty\)</span>,</p>
<div class="math notranslate nohighlight">
\[
  f_{d} \rightarrow f_{0} \prod_{j=1}^{d}  \frac{1+\beta /(j(1-\beta))}{1+2 /(j(1-\beta))}
  \]</div>
<p>which behave like <span class="math notranslate nohighlight">\(d ^{-\alpha}\)</span> for <span class="math notranslate nohighlight">\(\alpha = \frac{2-\beta}{1-\beta}\)</span>.</p>
</li>
<li><p>generates many more dense bipartite subgraphs than are found in comparable classical random graphs, which is encountered in real web graphs.</p></li>
</ul>
</div>
</div>
<div class="section" id="fitting">
<h3>Fitting<a class="headerlink" href="#fitting" title="Permalink to this headline">¶</a></h3>
<p>We need to observe how the graph change over time, i.e. a sequence of ‘snap-shots’ of the graph. But often we only have the final shot. In this situation, we can still do fitting. See [SAND 407] which makes clever usage of their recursive nature to get around this lack of multiple snap-shots.</p>
</div>
</div>
<div class="section" id="exponential-random-graph-models">
<h2>Exponential Random Graph Models<a class="headerlink" href="#exponential-random-graph-models" title="Permalink to this headline">¶</a></h2>
<p>ERGM are better than the above models, in construction, estimation, comparison etc.</p>
<p>Consider a random graph <span class="math notranslate nohighlight">\(G = (V, E)\)</span>, let <span class="math notranslate nohighlight">\(\boldsymbol{Y}\)</span> be the random adjacency matrix. An ERGM is a model specified in exponential family form for the joint distribution of the elements <span class="math notranslate nohighlight">\(y_{ij}\)</span>:</p>
<div class="math notranslate nohighlight">
\[
\mathbb{P} _\theta (\boldsymbol{Y} = \boldsymbol{y} ) = \frac{1}{\kappa} \exp \left\{ \sum_H \theta_H \cdot g_H(\boldsymbol{y}) \right\}
\]</div>
<p>where</p>
<ul class="simple">
<li><p>each <span class="math notranslate nohighlight">\(H\)</span> is a <strong>configuration</strong>: a set of possible edges among a subset of the vertices in <span class="math notranslate nohighlight">\(G\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(g_{H}(\boldsymbol{y})=\prod_{y_{i j} \in H} y_{i j}\)</span>, which is <span class="math notranslate nohighlight">\(1\)</span> if the configuration <span class="math notranslate nohighlight">\(H\)</span> occurs in <span class="math notranslate nohighlight">\(\boldsymbol{y}\)</span>, or <span class="math notranslate nohighlight">\(0\)</span> otherwise.</p></li>
<li><p>a non-zero value for <span class="math notranslate nohighlight">\(\theta_H\)</span> means that the <span class="math notranslate nohighlight">\(Y_{ij}\)</span> are independent in <span class="math notranslate nohighlight">\(H\)</span>, conditional upon the rest of the graph</p></li>
<li><p><span class="math notranslate nohighlight">\(\kappa = \kappa(\theta)=\sum_{\boldsymbol{y}} \exp \left\{\sum_{H} \theta_{H} g_{H}(\boldsymbol{y})\right\}\)</span> is a normalization constant.</p></li>
</ul>
<p>Note that</p>
<ul>
<li><p>The summation implies a certain (in)dependency structure among <span class="math notranslate nohighlight">\(Y_{ij}\)</span>. For given index set <span class="math notranslate nohighlight">\(\mathcal{A}, \mathcal{B}, \mathcal{C}\)</span>, the random variables <span class="math notranslate nohighlight">\(\left\{ Y_{i, j} \right\}_{(i, j) \in \mathcal{A}}\)</span> and independent of <span class="math notranslate nohighlight">\(\left\{ Y_{i, j} \right\}_{(i, j) \in \mathcal{B}}\)</span>, given the values of <span class="math notranslate nohighlight">\(\left\{ Y_{i, j} \right\}_{(i, j) \in \mathcal{C}}\)</span>.</p></li>
<li><p>Conversely, we can begin with a collection of (in)dependence relations among subsets of elements in <span class="math notranslate nohighlight">\(\boldsymbol{Y}\)</span> and try to develop a model. But certain conditions need to be satisfied, that are formalized in the Hammersley-Clifford theorem [SAND 36].</p></li>
<li><p>One nice property:</p>
<div class="math notranslate nohighlight">
\[
  \log \left[\frac{\mathbb{P}_{\theta}\left(Y_{i j}=1 \mid \boldsymbol{Y}_{(-i j)}=\boldsymbol{y}_{(-i j)}\right)}{\mathbb{P}_{\theta}\left(Y_{i j}=0 \mid \boldsymbol{Y}_{(-i j)}=\boldsymbol{y}_{(-i j)}\right)}\right]=\theta^{\top} \Delta_{i j}(\boldsymbol{y})
  \]</div>
</li>
</ul>
<div class="section" id="bernoulli-random-graphs">
<h3>Bernoulli Random Graphs<a class="headerlink" href="#bernoulli-random-graphs" title="Permalink to this headline">¶</a></h3>
<p>If we assume each edge <span class="math notranslate nohighlight">\(e(i, j)\)</span> is formed independently with probability <span class="math notranslate nohighlight">\(p_{ij}\)</span>, i.e. <span class="math notranslate nohighlight">\(y_{ij} \sim \operatorname{Ber}(p_{ij})\)</span> and <span class="math notranslate nohighlight">\(y_{i,j} \perp y_{i ^\prime , j ^\prime}\)</span> for any <span class="math notranslate nohighlight">\((i, j)\ne (i ^\prime , j ^\prime)\)</span>, then</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\theta_H = 0\)</span> for all configurations <span class="math notranslate nohighlight">\(H\)</span> involving three or more vertices.</p></li>
<li><p><span class="math notranslate nohighlight">\(g_H(\boldsymbol{y}) = g_{ij}(\boldsymbol{y}) = y_{ij}\)</span>.</p></li>
</ul>
<p>The ERGM model reduces to</p>
<div class="math notranslate nohighlight">
\[
\mathbb{P}_{\theta}(\boldsymbol{Y}=\boldsymbol{y})=\left(\frac{1}{\kappa} \right) \exp \left\{\sum_{i, j} \theta_{i j} y_{i j}\right\}
\]</div>
<p>which is another way of writing <span class="math notranslate nohighlight">\(p_{ij} = \frac{\exp(\theta_{ij})}{1+\exp(\theta_{ij})}\)</span>.</p>
<p>Obviously, this is <span class="math notranslate nohighlight">\(\mathcal{O} (N_v^2)\)</span> number of parameters. It is common to impose an assumption of homogeneity across vertex pairs, e.g. <span class="math notranslate nohighlight">\(\theta_{ij} \equiv \theta\)</span> for all <span class="math notranslate nohighlight">\((i, j)\)</span>. Hence</p>
<div class="math notranslate nohighlight">
\[
\mathbb{P}_{\theta}(\boldsymbol{Y}=\boldsymbol{y})=\left(\frac{1}{\kappa} \right) \exp \left\{\theta\sum_{i, j}  y_{i j}\right\}=\left(\frac{1}{\kappa} \right) \exp \left\{\theta N_e\right\}
\]</div>
<p>which is exactly the random graph model <span class="math notranslate nohighlight">\(\mathcal{G} (N_v, p)\)</span> with <span class="math notranslate nohighlight">\(p = \frac{\exp(\theta)}{\exp(\theta)}\)</span>.</p>
<p>More generally, we can consider a partition of two sets <span class="math notranslate nohighlight">\((S_1, S_2)\)</span> of vertices, and impose homogeneity within and between sets, i.e. 3 kinds of <span class="math notranslate nohighlight">\(\theta\)</span> values. The model is then</p>
<div class="math notranslate nohighlight">
\[
\mathbb{P}_{\theta}(\boldsymbol{Y}=\boldsymbol{y})=\left(\frac{1}{\kappa} \right) \exp \left\{\theta_{11}L_{11}(\boldsymbol{y})+ \theta_{12}L_{12}(\boldsymbol{y})+ \theta_{22}L_{22}(\boldsymbol{y})\right\}
\]</div>
<p>where <span class="math notranslate nohighlight">\(L_{11}(\boldsymbol{y})\)</span> and <span class="math notranslate nohighlight">\(L_{22}(\boldsymbol{y})\)</span> are the number within-set edges for <span class="math notranslate nohighlight">\(S_1\)</span> and <span class="math notranslate nohighlight">\(S_2\)</span> respectively, and <span class="math notranslate nohighlight">\(L_{12}(\boldsymbol{y})\)</span> is the number of across-set edges.</p>
<p>Cons</p>
<ul class="simple">
<li><p>assumption of complete independence is untenable in practice</p></li>
<li><p>Bernoulli-like random graphs lack many characteristics of real-world graphs</p></li>
</ul>
</div>
<div class="section" id="markov-random-graphs">
<h3>Markov Random Graphs<a class="headerlink" href="#markov-random-graphs" title="Permalink to this headline">¶</a></h3>
<p>[SAND 155]</p>
<p>Markov dependence: two possible edges are dependent whenever they share a vertex, conditional on all other possible edges. That is, the presence or absence of <span class="math notranslate nohighlight">\((i, j)\)</span> depends on that of <span class="math notranslate nohighlight">\((i, k)\)</span> for <span class="math notranslate nohighlight">\(k\ne j\)</span>, given information on all other edges. A random graph G arising under Markov dependence conditions is called a Markov graph.</p>
<p>What’s the joint distribution? Using the Hammersley-Clifford theorem, <span class="math notranslate nohighlight">\(G\)</span> is a Markov graph if and only if the joint distribution can be expressed as</p>
<div class="math notranslate nohighlight">
\[
\mathbb{P}_{\theta}(\boldsymbol{Y}=\boldsymbol{y})=\left(\frac{1}{\kappa} \right) \exp \left\{\sum_{k=1}^{N_{v}-1} \theta_{k} S_{k}(\boldsymbol{y})+\theta_{\tau} T(\boldsymbol{y})\right\}
\]</div>
<p>where</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(S_1(\boldsymbol{y} )=N_e\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(S_k(\boldsymbol{y})\)</span> is the number of <span class="math notranslate nohighlight">\(k\)</span>-stars (a tree with one vertex of degree <span class="math notranslate nohighlight">\(k\)</span> and <span class="math notranslate nohighlight">\(k\)</span> neighbors of degree 1).</p></li>
<li><p><span class="math notranslate nohighlight">\(T(\boldsymbol{y})\)</span> is the number of triangles</p></li>
</ul>
<p>Note that</p>
<ul class="simple">
<li><p>the statistics <span class="math notranslate nohighlight">\(S_k\)</span> and <span class="math notranslate nohighlight">\(S_{k ^\prime}\)</span> are expected to be correlated (<span class="math notranslate nohighlight">\(k\)</span> stars implies <span class="math notranslate nohighlight">\(k ^\prime\)</span>-stars for <span class="math notranslate nohighlight">\(k ^\prime &lt; k\)</span>)</p></li>
<li><p><span class="math notranslate nohighlight">\(S_k\)</span> and <span class="math notranslate nohighlight">\(T\)</span> are also correlated</p></li>
<li><p>fitting higher order of <span class="math notranslate nohighlight">\(k\)</span> is hard</p></li>
</ul>
</div>
<div class="section" id="id1">
<h3>Fitting<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>The MLEs <span class="math notranslate nohighlight">\(\hat{\theta}_H\)</span> of <span class="math notranslate nohighlight">\(\theta_H\)</span> are well defined but the calculation is non-trivial. There is no developed theory for CI, testing, due to the highly dependent nature of observations.</p>
<p>The log-likelihood is</p>
<div class="math notranslate nohighlight">
\[
\ell(\theta)=\theta^{\top} \boldsymbol{g}(\boldsymbol{y})-\psi(\theta)
\]</div>
<p>where <span class="math notranslate nohighlight">\(\psi(\theta) = \log \kappa (\theta)\)</span> and <span class="math notranslate nohighlight">\(\kappa(\theta)=\sum_{\boldsymbol{y}} \exp \left\{\theta^{\top} \boldsymbol{g}(\boldsymbol{y})\right\}\)</span>. Taking derivative w.r.t. <span class="math notranslate nohighlight">\(\theta\)</span> on both sides gives</p>
<div class="math notranslate nohighlight">
\[
\boldsymbol{g}(\boldsymbol{y}) = \frac{\partial \psi(\theta)}{\partial \theta}
\]</div>
<p>Note the fact <span class="math notranslate nohighlight">\(\mathbb{E}_{\theta}[\boldsymbol{g}(\boldsymbol{Y})]=\partial \psi(\theta) / \partial \theta\)</span>. Hence the MLE can be expressed as the solution to the system of equations</p>
<div class="math notranslate nohighlight">
\[
\boldsymbol{g}(\boldsymbol{y}) = \mathbb{E}_{\theta}[\boldsymbol{g}(\boldsymbol{Y})]
\]</div>
<div class="section" id="approximate-log-likelihood">
<h4>Approximate Log-likelihood<a class="headerlink" href="#approximate-log-likelihood" title="Permalink to this headline">¶</a></h4>
<p>However, <span class="math notranslate nohighlight">\(\psi(\theta)\)</span> is hard to evaluate since it involves <span class="math notranslate nohighlight">\(\sum_{\boldsymbol{y}}\)</span> over <span class="math notranslate nohighlight">\(2^{\binom{N_v}{2} }\)</span> possible choice of <span class="math notranslate nohighlight">\(\boldsymbol{y}\)</span>, for each <span class="math notranslate nohighlight">\(\theta\)</span>. Two Monte Carlo approaches are used</p>
<ul class="simple">
<li><p>stochastic approximation of the log-likelihood <span class="math notranslate nohighlight">\(\ell(\boldsymbol{\theta} )\)</span> [SAND 209]</p></li>
<li><p>stochastic approximation to solutions of systems of equations <span class="math notranslate nohighlight">\(\boldsymbol{g}(\boldsymbol{y}) = \frac{\partial \psi(\theta)}{\partial \theta}\)</span> by Robbins-Monro algorithm [SAND 324].</p></li>
</ul>
</div>
<div class="section" id="log-pseudo-likelihood">
<h4>Log-pseudo-likelihood<a class="headerlink" href="#log-pseudo-likelihood" title="Permalink to this headline">¶</a></h4>
<p>A disadvantage of both of the methods above is their computationally intensive nature. To date they have been applied to networks with at most a few thousand vertices. An alternative is to estimate <span class="math notranslate nohighlight">\(\theta\)</span> by maximizing the log-pseudo-likelihood</p>
<div class="math notranslate nohighlight">
\[
\sum_{\{i, j\}} \log \mathbb{P}_{\theta}\left(Y_{i j}=1 \mid \boldsymbol{Y}_{(-i j)}=\boldsymbol{y}_{(-i j)}\right)
\]</div>
<p>pros</p>
<ul class="simple">
<li><p>work best when dependencies among the elements of <span class="math notranslate nohighlight">\(\boldsymbol{y}\)</span> are relatively weak.</p></li>
<li><p>computationally expedient</p></li>
</ul>
<p>cons</p>
<ul class="simple">
<li><p>the above condition does not hold in many real-world contexts</p></li>
</ul>
<p>See [SAND 37, 372]</p>
</div>
</div>
<div class="section" id="goodness-of-fit">
<h3>Goodness-of-fit<a class="headerlink" href="#goodness-of-fit" title="Permalink to this headline">¶</a></h3>
<p>Still not well developed for graph modeling contexts. Some practice steps are</p>
<ul class="simple">
<li><p>simulate numerous random graphs from the fitted model</p></li>
<li><p>compare high-level summary characteristics of these graphs, with those of the originally observed graph</p></li>
<li><p>if poor match, then this suggests systematic differences between the specified class of models and the data, i.e. lack of goodness-of-fit</p></li>
</ul>
</div>
<div class="section" id="model-degeneracy">
<h3>Model Degeneracy<a class="headerlink" href="#model-degeneracy" title="Permalink to this headline">¶</a></h3>
<dl class="simple myst">
<dt>Definition (Model degeneracy)</dt><dd><p>a probability distribution that places a disproportionately large amount of its mass on a correspondingly small set of outcomes.</p>
<ul class="simple">
<li><p>For instance, a number of simple but popular Markov graph models have been shown to be degenerate: places most of its mass on either the empty graph, the complete graph, or a mixture of the two, depending on the value of <span class="math notranslate nohighlight">\(\theta\)</span>.</p></li>
</ul>
</dd>
</dl>
<p>For degenerate models, the transition boundary (from empty, to mixture, to complete) of <span class="math notranslate nohighlight">\(\theta\)</span> is sharp. As a result, for poorly specified ERGMs, numerical algorithms like MCMC can have substantial difficulty converging.</p>
<p>For a case study see [SAND pg.188].</p>
</div>
</div>
<div class="section" id="challenges">
<h2>Challenges<a class="headerlink" href="#challenges" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>How to incorporate our understanding of working of an underlying complex system into the model?</p></li>
<li><p>inference</p></li>
<li><p>model specification</p></li>
</ul>
<p>Related topics</p>
<ul class="simple">
<li><p>dynamical/complex systems</p></li>
<li><p>agent-based modeling (economics, sociology)</p></li>
<li><p>graphical models</p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./38-ml-for-graph-data"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="13-sampling-and-estimation.html" title="previous page">Sampling and Estimation</a>
    <a class='right-next' id="next-link" href="31-topology-inference.html" title="next page">Topology Inference</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Dennis Zheng<br/>
        
            &copy; Copyright 2020.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
<script async="" src="https://www.google-analytics.com/analytics.js"></script>
<script>
                        window.ga = window.ga || function () {
                            (ga.q = ga.q || []).push(arguments) };
                        ga.l = +new Date;
                        ga('create', 'UA-150740237-2', 'auto');
                        ga('set', 'anonymizeIp', true);
                        ga('send', 'pageview');
                    </script>

  </body>
</html>